### PR TITLE
Mypy Checking to pathod

### PR DESCRIPTION
**Pull request by iharsh234**
This PR Extend Mypy checking to `/pathod` folder , fixed some bugs from previous` example/simple `and add these mypy checking to `tox`.
REF #2194 , #2221

### CODE REVIEW

File: pathod/language/websockets.py [hunk 1]
```diff
@@ -239,8 +239,14 @@ class NestedFrame(base.NestedMessage):
     preamble = "f"
     nest_type = WebsocketFrame
 
-COMP = Tuple[ Type[OpCode], Type[Length], Type[Fin], Type[RSV1], Type[RSV2], Type[RSV3], Type[Mask], Type[actions.PauseAt], Type[actions.DisconnectAt], Type[actions.InjectAt], Type[KeyNone], Type[Key], Type[Times], Type[Body], Type[RawBody]]
+
+COMP = Tuple[
```
Comment (iharsh234) [inline pathod/language/websockets.py]: @Kriechi Any better idea how to fix this , or ignoring type is the only choice?

**Comment by cortesi** [comment]
Hi there. This looks good to me. One point about code style - as a general rule, we import modules, and not individual items from modules. So we prefer:

```python
import typing

x = []  # type: typing.Sequence
```

to

```python
from typing import Sequence

x = []  # type: Sequence
```

This is a bit more verbose, but is a lot clearer, doesn't require you to constantly adjust the imports as you use more or less of the module, and really pays dividends when refactoring or doing searches across the codebase.

The particular bit of code you're working on here doesn't stick to the import rules very consistently - feel free to fix the usage if you're working on the code anyway.

**Comment by iharsh234** [comment]
@cortesi done :smile:

File: pathod/language/actions.py [hunk 2]
```diff
@@ -52,7 +50,7 @@ def intermediate(self, settings):  # pragma: no cover
 
 
 class PauseAt(_Action):
-    unique_name = None
+    unique_name = None  # type: ignore
```
Comment (cortesi) [inline pathod/language/actions.py]: This can be `typing.Optional[str]`. Same for all the similar occurrences.

Comment (iharsh234) [inline pathod/language/actions.py]: I have done that before , but was getting some errors like
`error: Definition of "unique_name" in base class "_HeaderMixin" is incompatible with definition in base class "Token"
`
Then from [here](https://github.com/python/mypy/issues/1237) I get to know that either we should refactor code or go for `# type: ignore`

Comment (cortesi) [inline pathod/language/actions.py]: Ah, OK then. Squash the commits and let's merge this puppy.

File: examples/simple/custom_contentview.py [hunk 3]
```diff
@@ -14,7 +17,7 @@ class ViewSwapCase(contentviews.View):
     prompt = ("swap case text", "z")
     content_types = ["text/plain"]
 
-    def __call__(self, data: bytes, **metadata) -> Tuple[str, Iterable[List[Tuple[str, AnyStr]]]]:
+    def __call__(self, data: typing.AnyStr, **metadata) -> CVIEWSWAPCASE:
```
Comment (mhils) [inline examples/simple/custom_contentview.py]: Why `AnyStr` here? This should be `bytes`.

File: examples/simple/io_read_dumpfile.py [hunk 4]
```diff
@@ -1,8 +1,9 @@
 #!/usr/bin/env python
+
+# type: ignore
```
Comment (mhils) [inline examples/simple/io_read_dumpfile.py]: What's the motivation for this?

File: examples/simple/io_write_dumpfile.py [hunk 5]
```diff
@@ -8,12 +8,13 @@
...
 import sys
 from mitmproxy import io, http
+import typing  # noqa
 
 
 class Writer:
     def __init__(self, path: str) -> None:
         if path == "-":
-            f = sys.stdout  # type: io.TextIO
+            f = sys.stdout  # type: typing.IO[typing.Any]
```
Comment (mhils) [inline examples/simple/io_write_dumpfile.py]: The FlowWriter only writes bytes, so this should really be `typing.IO[bytes]`. If that raises errors, that's a strong indicator that the code here does not work as intended. I'd be fine with removing the `if path == "-":` special case altogether.

File: pathod/language/actions.py (hunk 2 above)
Comment (mhils) [inline pathod/language/actions.py]: That's not really what the linked issue says. These attributes should have the same type as the base class.

File: pathod/language/base.py [hunk 6]
```diff
@@ -376,7 +375,7 @@ def freeze(self, settings):
 
 
 class Integer(_Component):
-    bounds = (None, None)
+    bounds = (None, None)  # type: typing.Tuple[typing.Union[int, None], typing.Union[int , None]]
```
Comment (mhils) [inline pathod/language/base.py]: This should be `Optional[int]`, not `typing.Union[int, None]`.

File: pathod/language/base.py [hunk 7]
```diff
@@ -546,7 +545,7 @@ class NestedMessage(Token):
         A nested message, as an escaped string with a preamble.
     """
     preamble = ""
-    nest_type = None
+    nest_type = None  # type: ignore
```
Comment (mhils) [inline pathod/language/base.py]: We shouldn't ignore this here either.

File: pathod/language/websockets.py [hunk 8]
```diff
@@ -239,7 +240,14 @@ class NestedFrame(base.NestedMessage):
     nest_type = WebsocketFrame
 
 
+COMP = typing.Tuple[
+    typing.Type[OpCode], typing.Type[Length], typing.Type[Fin], typing.Type[RSV1], typing.Type[RSV2], typing.Type[RSV3], typing.Type[Mask],
+    typing.Type[actions.PauseAt], typing.Type[actions.DisconnectAt], typing.Type[actions.InjectAt], typing.Type[KeyNone], typing.Type[Key],
+    typing.Type[Times], typing.Type[Body], typing.Type[RawBody]
+]
```
Comment (mhils) [inline pathod/language/websockets.py]: That looks wrong. First, this doesn't really match the semantics of [`typing.Tuple`](https://docs.python.org/3/library/typing.html#typing.Tuple), second, I think all those are `_Component`s?

File: pathod/pathod.py [hunk 9]
```diff
@@ -71,7 +69,7 @@ def get_cert(self, name):
 
 class PathodHandler(tcp.BaseHandler):
     wbufsize = 0
-    sni = None
+    sni = None  # type: typing.Union[str, None, bool]
```
Comment (mhils) [inline pathod/pathod.py]: Why would this be a bool? Looks like `Optional[str]` to me?

File: examples/simple/custom_contentview.py [hunk 10]
```diff
@@ -3,7 +3,10 @@
 The content view API is explained in the mitmproxy.contentviews module.
 """
 from mitmproxy import contentviews
-from typing import Tuple, Iterable, AnyStr, List
+import typing
+
+
+CVIEWSWAPCASE = typing.Tuple[str, typing.Iterable[typing.List[typing.Tuple[str, typing.AnyStr]]]]
```
Comment (mhils) [inline examples/simple/custom_contentview.py]: This type annotations should be done here: https://github.com/mitmproxy/mitmproxy/blob/9d2350b6709532bf185c6e0f56342b7b780092c7/mitmproxy/contentviews/base.py#L17 and then re-used here. Please prefix types with `T`.

File: examples/simple/io_read_dumpfile.py (hunk 4 above)
Comment (iharsh234) [inline examples/simple/io_read_dumpfile.py]: It was previous fix , but after adding `--follow-imports=skip` , it should not be needed now. thanks:)

File: pathod/language/base.py (hunk 6 above)
Comment (iharsh234) [inline pathod/language/base.py]: Don't both of them same?

File: pathod/language/websockets.py (hunk 8 above)
Comment (iharsh234) [inline pathod/language/websockets.py]: I wasn't sure about it , and asked but didn't received any reply.

File: pathod/language/actions.py (hunk 2 above)
Comment (iharsh234) [inline pathod/language/actions.py]: They should have , but mypy always says they are incompatible , don't know why

File: pathod/pathod.py (hunk 9 above)
Comment (iharsh234) [inline pathod/pathod.py]: > sni: ``str`` for a custom SNI value,
                ``None`` for the client SNI value,
                ``False`` if no SNI value should be sent.

File: examples/simple/custom_contentview.py (hunk 3 above)
Comment (iharsh234) [inline examples/simple/custom_contentview.py]: indeed it should be, it working with mypy test now.

File: examples/simple/io_write_dumpfile.py (hunk 5 above)
Comment (iharsh234) [inline examples/simple/io_write_dumpfile.py]:
```
def __init__(self, path: str) -> None:
        f = open(path, "wb")  # type: typing.IO[bytes]
        self.w = io.FlowWriter(f)
```
this worked as expected :smile:

**Pull request merged**